wallet/invoices.c

```diff
@@ -82,20 +86,112 @@ struct invoices *invoices_new(const tal_t *ctx,
 	invs->db = db;
 	invs->log = log;
 	invs->timers = timers;
+	invs->expiry_trigger = NULL;
+	invs->min_expiry_time = 0;
 
 	list_head_init(&invs->invlist);
 	list_head_init(&invs->waitany_waiters);
 
 	return invs;
 }
 
+static void setup_expiry_trigger(struct invoices *invoices, u64 now);
+static void on_expiry_trigger(struct invoices *invoices)
+{
+	u64 now = time_now().ts.tv_sec;
+	struct invoice *i;
+	struct invoice_waiter *w;
+	const tal_t *tmpctx;
+	sqlite3_stmt *stmt;
+
+	tmpctx = tal_tmpctx(invoices);
+	list_for_each(&invoices->invlist, i, list) {
```

Comment:
OK, so our timers are scalable.  Like, we don't worry for the first million or so!

So this is better done with per-invoice timers, rather than scanning like this.

Reply:
The only subtlety is time_mono vs time_abs; so an invoice expiry timer needs to check if it really is expired, and re-arm itself if it's still got some time to go.  (It's OK if it misses a few seconds the other way, IMHO).

Reply:
Also, please note in the commit message that this changes the JSON API for waitinvoice!

Reply:
Oh, ok.  I shall do so later.